src/hotspot/cpu/aarch64/templateTable_aarch64.cpp

```diff
@@ -3242,54 +3292,24 @@ void TemplateTable::fast_xaccess(TosState state)
 //-----------------------------------------------------------------------------
 // Calls
 
-void TemplateTable::prepare_invoke(int byte_no,
-                                   Register method, // linked method (or i-klass)
-                                   Register index,  // itable index, MethodType, etc.
-                                   Register recv,   // if caller wants to see it
+void TemplateTable::prepare_invoke(Register recv,   // if caller wants to see it
                                    Register flags   // if caller wants to test it
                                    ) {
-  // determine flags
+
+  const Register cache = r2;
   Bytecodes::Code code = bytecode();
-  const bool is_invokeinterface  = code == Bytecodes::_invokeinterface;
-  const bool is_invokedynamic    = code == Bytecodes::_invokedynamic;
-  const bool is_invokehandle     = code == Bytecodes::_invokehandle;
-  const bool is_invokevirtual    = code == Bytecodes::_invokevirtual;
-  const bool is_invokespecial    = code == Bytecodes::_invokespecial;
-  const bool load_receiver       = (recv  != noreg);
-  const bool save_flags          = (flags != noreg);
-  assert(load_receiver == (code != Bytecodes::_invokestatic && code != Bytecodes::_invokedynamic), "");
-  assert(save_flags    == (is_invokeinterface || is_invokevirtual), "need flags for vfinal");
-  assert(flags == noreg || flags == r3, "");
-  assert(recv  == noreg || recv  == r2, "");
-
-  // setup registers & access constant pool cache
-  if (recv  == noreg)  recv  = r2;
-  if (flags == noreg)  flags = r3;
-  assert_different_registers(method, index, recv, flags);
+  const bool load_receiver       = (code != Bytecodes::_invokestatic) && (code != Bytecodes::_invokedynamic);
+  assert_different_registers(recv, flags);
 
   // save 'interpreter return address'
-  __ save_bcp();
+  __ save_bcp(); // probably don't need this
 
-  load_invoke_cp_cache_entry(byte_no, method, index, flags, is_invokevirtual, false, is_invokedynamic);
-
-  // maybe push appendix to arguments (just before return address)
-  if (is_invokehandle) {
-    Label L_no_push;
-    __ tbz(flags, ConstantPoolCacheEntry::has_appendix_shift, L_no_push);
-    // Push the appendix as a trailing parameter.
-    // This must be done before we get the receiver,
-    // since the parameter_size includes it.
-    __ push(r19);
-    __ mov(r19, index);
-    __ load_resolved_reference_at_index(index, r19);
-    __ pop(r19);
-    __ push(index);  // push appendix (MethodType, CallSite, etc.)
-    __ bind(L_no_push);
-  }
+  // Load TOS state for later
+  __ load_unsigned_byte(rscratch2, Address(cache, in_bytes(ResolvedMethodEntry::type_offset())));
```

Comment:
In the old code this came after the `if (load_receiver)` block just before rscratch2 was used to compute the return address. This was for a good reason. At the end of the `if` block there is a call to `verify_oop` which uses rscratch2 when VerifyOops is set. So, relocating this load of the TOS state will fail to compute the correct return address in that case.

Reply:
Oops, ignore that! `verify_oops` pushes rscratch2 on the stack before using it and restores it afterwards.